[1624] Add the support Collapse/Expand of a node

gcoutable opened the pull request:
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

## Review

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

gcoutable committed:
[1624] Add the support for collapsing/expanding nodes

Bug: https://github.com/eclipse-sirius/sirius-components/issues/1624
Signed-off-by: Guillaume Coutable <guillaume.coutable@obeo.fr>

pcdavid commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/java/org/eclipse/sirius/components/collaborative/diagrams/dto/CollapseExpandDiagramElementSuccessPayload.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @author gcoutable
+ */
+public class CollapseExpandDiagramElementSuccessPayload implements IPayload {
```
This can now be replaced by the generic `SuccessPayload`. See https://github.com/eclipse-sirius/sirius-components/commit/23086a2757958949ed9984c45c0c2e97791a602e.

gcoutable replied:
Done

sbegaudeau commented on CHANGELOG.adoc:
```diff
@@ -76,6 +76,7 @@ image:doc/images/View_edit_begin_label_in_action.png[width=550,height=234]
 +
 image:doc/images/Free_form_layout_in_a_list_layout.png[A node with three free form compartment]
 
+- https://github.com/eclipse-sirius/sirius-components/issues/1624[#1624] [diagram] Add support for expand/collapse of nodes. A node can be collapsed or expanded if its view description is collapsible or if its sirius desktop description is a region (Horizontal or Vertical stack).
```
One line per sentence

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/java/org/eclipse/sirius/components/collaborative/diagrams/dto/CollapseExpandDiagramElementInput.java:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @author gcoutable
+ */
+public record CollapseExpandDiagramElementInput(UUID id, String editingContextId, String representationId, String elementId, CollapsingState collapsingState) implements IDiagramInput {
```
I'll rename that to match the new name in the ADR

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/java/org/eclipse/sirius/components/collaborative/diagrams/handlers/CollapseExpandDiagramElementEventHandler.java:
```diff
@@ -0,0 +1,103 @@
+ * @author gcoutable
+ */
+@Service
+public class CollapseExpandDiagramElementEventHandler implements IDiagramEventHandler {
```
I'll rename that to match the new name in the ADR

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/java/org/eclipse/sirius/components/collaborative/diagrams/handlers/CollapseExpandDiagramElementEventHandler.java:
```diff
@@ -0,0 +1,103 @@
+    public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDescriptionSink, IEditingContext editingContext, IDiagramContext diagramContext, IDiagramInput diagramInput) {
+        this.counter.increment();
+
+        if (diagramInput instanceof CollapseExpandDiagramElementInput) {
```
use the instanceof pattern matching to cast the variable directly

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -8,7 +8,10 @@ input DiagramEventInput {
 }
 
-union DiagramEventPayload = ErrorPayload | SubscribersUpdatedEventPayload | DiagramRefreshedEventPayload
+union DiagramEventPayload =
```
You seems to be using a tool which changes the formatting of the graphqls files, please be careful with that.

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -197,7 +200,10 @@ type DiagramDescription implements RepresentationDescription {
+  connectorTools(
+    sourceDiagramElementId: ID!
+    targetDiagramElementId: ID!
+  ): [Tool!]!
```
No need to change that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -229,15 +235,30 @@ extend type Mutation {
+  fadeDiagramElement(
+    input: FadeDiagramElementInput!
+  ): FadeDiagramElementPayload!
+  collapseExpandDiagramElement(
```
No need to change that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -229,15 +235,30 @@ extend type Mutation {
+    input: FadeDiagramElementInput!
+  ): FadeDiagramElementPayload!
+  collapseExpandDiagramElement(
+    input: CollapseExpandDiagramElementInput!
```
I'll rename that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -297,7 +318,9 @@ input InvokeSingleClickOnTwoDiagramElementsToolInput {
-union InvokeSingleClickOnTwoDiagramElementsToolPayload = ErrorPayload | InvokeSingleClickOnTwoDiagramElementsToolSuccessPayload
+union InvokeSingleClickOnTwoDiagramElementsToolPayload =
+    ErrorPayload
+  | InvokeSingleClickOnTwoDiagramElementsToolSuccessPayload
```
No need to change that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -315,7 +338,9 @@ input InvokeSingleClickOnDiagramElementToolInput {
-union InvokeSingleClickOnDiagramElementToolPayload = ErrorPayload | InvokeSingleClickOnDiagramElementToolSuccessPayload
+union InvokeSingleClickOnDiagramElementToolPayload =
+    ErrorPayload
+  | InvokeSingleClickOnDiagramElementToolSuccessPayload
```
No need to change that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -349,7 +374,9 @@ input UpdateNodePositionInput {
-union UpdateNodePositionPayload = ErrorPayload | UpdateNodePositionSuccessPayload
+union UpdateNodePositionPayload =
+    ErrorPayload
+  | UpdateNodePositionSuccessPayload
```
No need to change that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -423,5 +450,23 @@ input FadeDiagramElementInput {
   fade: Boolean!
 }
 
+union FadeDiagramElementPayload = ErrorPayload | SuccessPayload
```
No need to change that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/resources/schema/diagram.graphqls:
```diff
@@ -423,5 +450,23 @@ input FadeDiagramElementInput {
+  collapsingState: CollapsingState!
+}
+
+union CollapseExpandDiagramElementPayload = ErrorPayload | SuccessPayload
```
I'll rename that

sbegaudeau commented on packages/diagrams/backend/sirius-components-collaborative-diagrams/src/test/java/org/eclipse/sirius/components/collaborative/diagrams/handlers/CollapseExpandDiagramElementEventHandlerTests.java:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @author gcoutable
+ */
+public class CollapseExpandDiagramElementEventHandlerTests {
```
I'll rename that

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams-graphql/src/main/java/org/eclipse/sirius/components/diagrams/graphql/datafetchers/mutation/MutationExpandCollapseDiagramElementDataFetcher.java:
```diff
@@ -0,0 +1,59 @@
+ *
+ * @author gcoutable
+ */
+@MutationDataFetcher(type = "Mutation", field = "collapseExpandDiagramElement")
```
I'll rename that

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/NodeCollapseExpandTests.java:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @author gcoutable
+ */
+public class NodeCollapseExpandTests {
```
I'll rename that

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams-layout/src/main/java/org/eclipse/sirius/components/diagrams/layout/incremental/NodeContext.java:
```diff
@@ -136,6 +141,14 @@ public NodeContext(NodeLayoutData nodeLayoutData, ISiriusWebLayoutConfigurator l
         this.labelPlacements = nodeLabelPlacements;
+
+        boolean nodeIsBeingExpanded = false;
+        if (optionalDiagramEvent.isPresent() && optionalDiagramEvent.get() instanceof CollapseExpandDiagramElementEvent collapseExpandDiagramEvent) {
```
I'll rename that

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/services/DefaultTestDiagramDescriptionProvider.java:
```diff
@@ -285,11 +285,15 @@ private Function<VariableManager, List<org.eclipse.sirius.components.representat
+        return this.getDefaultDiagramDescription(diagram, false);
+    }
+
+    public DiagramDescription getDefaultDiagramDescription(Diagram diagram, boolean isCollapsible) {
```
Do not specify such low level details here. You can simply clone a new NodeDescription. You don't need to modify that

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/services/DefaultTestDiagramDescriptionProvider.java:
```diff
@@ -370,7 +374,7 @@ private EdgeDescription getDefaultEdgeDescription(Diagram diagram, LabelDescript
     }
 
-    private NodeDescription getDefaultNodeDescription(LabelDescription labelDescription, NodeDescription childDescription) {
+    private NodeDescription getDefaultNodeDescription(LabelDescription labelDescription, NodeDescription childDescription, boolean isCollapsible) {
```
No need to change that, simply clone the NodeDescription for your test, it's done this way on purpose

sbegaudeau replied:
Technically a piece was missing that I'll add to make such use cases more easy to code. It was only working on Node/Diagram and not on NodeDescription/DiagramDescription

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeComponent.java:
```diff
@@ -212,6 +223,23 @@ private Element doRender(VariableManager nodeVariableManager, String targetObjec
         return new Element(NodeElementProps.TYPE, nodeElementPropsBuilder.build());
     }
 
+    private CollapsingState shouldRenderChildren(String nodeId, Optional<Node> optionalPreviousNode, Optional<IDiagramEvent> optionalDiagramEvent) {
```
Let's be coherent and name this method something like `computeCollapsingState`. The fact that we will not render the children is only the current consequence

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeComponent.java:
```diff
@@ -187,7 +193,11 @@ private Element doRender(VariableManager nodeVariableManager, String targetObjec
-        Set<CustomizableProperties> customizableProperties = optionalPreviousNode.map(Node::getCustomizedProperties).orElse(Set.of());
+
+        Set<CustomizableProperties> customizableProperties = Set.of();
+        if (CollapsingState.EXPANDED.equals(collapsingState)) {
```
Does that mean that we will never keep the size of a manually resized collapsed node? Because I would like to resize my collapsed nodes 😛

sbegaudeau replied:
We will need to discuss about that, I'll just open an issue for now to track this

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/events/CollapseExpandDiagramElementEvent.java:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @author gcoutable
+ */
+public record CollapseExpandDiagramElementEvent(String elementId, CollapsingState collapsingState) implements IDiagramEvent {
```
I'll rename that

sbegaudeau commented on packages/diagrams/frontend/sirius-components-diagrams/src/palette/ContextualPalette.tsx:
```diff
@@ -222,6 +223,10 @@ export const ContextualPalette = ({
       } else if (tool.id === 'graphical-delete') {
         invokeDelete(GQLDeletionPolicy.GRAPHICAL);
+      } else if (tool.id === 'expand') {
+        invokeCollapseExpand(GQLCollapsingState.EXPANDED);
```
I'll rename that

sbegaudeau commented on packages/diagrams/frontend/sirius-components-diagrams/src/representation/DiagramRepresentation.tsx:
```diff
@@ -60,6 +60,10 @@ import {
   GQLArrangeAllData,
   GQLArrangeAllInput,
   GQLArrangeAllVariables,
+  GQLCollapseExpandDiagramElementData,
```
I'll rename that

sbegaudeau commented on packages/diagrams/frontend/sirius-components-diagrams/src/palette/ContextualPalette.types.ts:
```diff
@@ -24,6 +24,7 @@ export interface ContextualPaletteProps {
   invokeClose: () => void;
   invokeHide: () => void;
   invokeFade: () => void;
+  invokeCollapseExpand: (collapseingState: GQLCollapsingState) => void;
```
I'll rename that

sbegaudeau commented on packages/diagrams/frontend/sirius-components-diagrams/src/representation/DiagramRepresentation.types.ts:
```diff
@@ -565,3 +568,28 @@ export interface GQLFadeDiagramElementPayload {
+  COLLAPSED = 'COLLAPSED',
+}
+
+export interface GQLCollapseExpandDiagramElementVariables {
```
I'll rename that

sbegaudeau commented on packages/diagrams/frontend/sirius-components-diagrams/src/representation/operations.ts:
```diff
@@ -266,6 +266,20 @@ export const fadeDiagramElementMutation = gql`
   }
 `;
 
+export const collapseExpandDiagramElementMutation = gql`
```
I'll rename that

sbegaudeau commented on packages/view/backend/sirius-components-view-edit/src/main/resources/plugin.properties:
```diff
@@ -226,7 +227,7 @@ _UI_RadioDescription_style_feature = Style
 _UI_BarChartDescription_valuesExpression_feature = Values Expression
 _UI_BarChartDescription_keysExpression_feature = Keys Expression
-_UI_BarChartDescription_yAxisLabelExpression_feature = Y Axis Label Expression
+_UI_BarChartDescription_yAxisLabelExpression_feature = YAxis Label Expression
```
Be careful, some labels have been edited manually

sbegaudeau commented on packages/view/backend/sirius-components-view-edit/src/main/resources/plugin.properties:
```diff
@@ -242,7 +243,7 @@ _UI_ButtonDescription_body_feature = Body
 _UI_ButtonDescription_style_feature = Style
 _UI_ButtonDescription_conditionalStyles_feature = Conditional Styles
-_UI_ImageDescription_urlExpression_feature = URL Expression
+_UI_ImageDescription_urlExpression_feature = Url Expression
```
That one too

sbegaudeau reviewed (changes requested):
I still have to test it to see how it performs with Papaya and rename a couple of things but otherwise, the code looks good. I'll do that first thing in the morning tomorrow.

sbegaudeau commented on packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeComponent.java:
```diff
@@ -156,13 +158,17 @@ private Element doRender(VariableManager nodeVariableManager, String targetObjec
         String targetObjectLabel = nodeDescription.getTargetObjectLabelProvider().apply(nodeVariableManager);
         Set<ViewModifier> modifiers = this.computeModifiers(optionalDiagramEvent, optionalPreviousNode, nodeId);
         ViewModifier state = this.computeState(modifiers);
+        CollapsingState collapsingState = this.shouldRenderChildren(nodeId, optionalPreviousNode, optionalDiagramEvent);
```
The style provider does not seems to have access to the `collapsingState`, I'll update that and provide an example in Papaya

sbegaudeau committed:
[1624] Make the collapsing state available as a variable to compute the style

This commit also updates Papaya to leverage this variable for operational actors

Bug: https://github.com/eclipse-sirius/sirius-components/issues/1624
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>
